diff --git configs/zynq_zybo_defconfig configs/zynq_zybo_defconfig
index 7c23fec..8fc954e 100644
--- configs/zynq_zybo_defconfig
+++ configs/zynq_zybo_defconfig
@@ -22,3 +22,5 @@ CONFIG_DEBUG_UART_ZYNQ=y
 CONFIG_DEBUG_UART_BASE=0xe0001000
 CONFIG_DEBUG_UART_CLOCK=50000000
 CONFIG_ZYNQ_QSPI=y
+# CONFIG_OF_SEPARATE is not set
+CONFIG_OF_EMBED=y
diff --git include/configs/zynq-common.h include/configs/zynq-common.h
index 982905d..4c3990c 100644
--- include/configs/zynq-common.h
+++ include/configs/zynq-common.h
@@ -227,10 +227,35 @@
 	"usbboot=if usb start; then " \
 			"echo Copying FIT from USB to RAM... && " \
 			"load usb 0 ${load_addr} ${fit_image} && " \
-			"bootm ${load_addr}; fi\0" \
-		DFU_ALT_INFO
-
-#define CONFIG_BOOTCOMMAND		"run $modeboot"
+			"bootm ${load_addr}\0" \
+		"fi\0" \
+	"bootenv=uEnv.txt\0" \
+        "model=" CONFIG_DEFAULT_DEVICE_TREE "\0" \
+	"loadbootenv=load mmc 0 ${load_addr} ${bootenv}\0" \
+	"importbootenv=echo Importing environment from mmc ...;" \
+		"env import -t $load_addr $filesize\0" \
+        "loadbootscript=load mmc 0 ${load_addr} boot.scr\0" \
+        "bootscript=echo Running bootscript from mmc ...;" \
+                "source ${load_addr}\0" \
+	DFU_ALT_INFO
+
+#define CONFIG_BOOTCOMMAND \
+	"if mmc rescan; then " \
+		"echo SD/MMC found on device...;" \
+		"if run loadbootenv; then " \
+			"echo Loaded environment from ${bootenv};" \
+			"run importbootenv;" \
+		"fi;" \
+		"if test -n $uenvcmd; then " \
+			"echo Running uenvcmd ...;" \
+			"run uenvcmd;" \
+		"fi;" \
+		"if run loadbootscript; then " \
+			"run bootscript; " \
+		"fi; " \
+	"fi;" \
+        "run $modeboot"
+#define CONFIG_CMD_BOOTZ
 #define CONFIG_BOOTDELAY		3 /* -1 to Disable autoboot */
 #define CONFIG_SYS_LOAD_ADDR		0 /* default? */
 
